pallets/gear-builtin/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+
+    fn source(&self) -> ProgramId;
+    fn destination(&self) -> BuiltinId;
+    fn payload_bytes(&self) -> &[u8];
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct FromStoredDispatch<T: Config> {
+    source: ProgramId,
+    destination: BuiltinId,
+    payload: Vec<u8>,
+    _phantom: sp_std::marker::PhantomData<T>,
+}
+
+impl<T: Config> Dispatchable for FromStoredDispatch<T> {
+    type Payload = Vec<u8>;
+
+    fn new(source: ProgramId, destination: BuiltinId, payload: Vec<u8>) -> Self {
+        Self {
+            source,
+            destination,
+            payload,
+            _phantom: Default::default(),
+        }
+    }
+
+    fn source(&self) -> ProgramId {
+        self.source
+    }
+
+    fn destination(&self) -> BuiltinId {
+        self.destination
+    }
+
+    fn payload_bytes(&self) -> &[u8] {
+        self.payload.as_ref()
+    }
+}
+
+impl<T: Config> TryFrom<StoredDispatch> for FromStoredDispatch<T> {
+    type Error = BuiltinActorError;
+
+    fn try_from(dispatch: StoredDispatch) -> Result<Self, Self::Error> {
+        let builtin_id =
+            <Pallet<T> as BuiltinRouterProvider<StoredDispatch, JournalNote, u64>>::provide()
+                .lookup(&dispatch.destination())
+                .ok_or(BuiltinActorError::UnknownActor)?;
+        Ok(Self {
+            source: dispatch.source(),
+            destination: builtin_id,
+            payload: dispatch.payload_bytes().to_vec(),
+            _phantom: Default::default(),
+        })
+    }
+}
+
+/// Built-in actor error type
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::Display)]
+pub enum BuiltinActorError {
+    /// Occurs if the underlying call has the weight greater than the `gas_limit`.
+    #[display(fmt = "Not enough gas supplied")]
+    InsufficientGas,
+    /// Occurs if the dispatch's message can't be decoded into a known type.
+    #[display(fmt = "Failure to decode message")]
+    UnknownMessageType,
+    /// Indicates an unreachable (under normal conditions) state:
+    /// - the `BuiltinId` passed to the `handle()` function does not match any known actors.
+    #[display(fmt = "Unknown builtin id passed to the `handle()` method")]
+    UnknownActor,
+}
+
+impl From<BuiltinActorError> for ActorExecutionErrorReplyReason {
+    /// Convert [`BuiltinActorError`] to [`core_processor::common::ActorExecutionErrorReplyReason`]
+    fn from(err: BuiltinActorError) -> Self {
+        match err {
+            BuiltinActorError::InsufficientGas => {
+                ActorExecutionErrorReplyReason::Trap(TrapExplanation::GasLimitExceeded)
+            }
+            BuiltinActorError::UnknownMessageType => ActorExecutionErrorReplyReason::Trap(
+                TrapExplanation::Panic("Message decoding error".to_string().into()),
+            ),
+            // This should convey a message of hitting an unreachable state
+            BuiltinActorError::UnknownActor => {
```

Comment:
Another update: now we actually do need a value as the output of the `lookup` in order to enable the check whether a dispatch can be handled by a builtin actor before consuming it (to avoid excessive cloning of a dispatch as discussed above in reference to the `queue.rs` from the `pallet_gear`).